pkg/plugin/sdk/destination.go

```diff
@@ -0,0 +1,249 @@
+	Configure(context.Context, map[string]string) error
+
+	// Open is called after Configure to signal the plugin it can prepare to
+	// start writing records. If needed, the plugin should open connections in
+	// this function.
+	Open(context.Context) error
+
+	// WriteAsync receives a Record and can cache it internally and write it at
+	// a later time. Once the record is successfully written to the destination
+	// the plugin must call the provided AckFunc function with a nil error. If
+	// the plugin failed to write the record to the destination it must call the
+	// supplied AckFunc with a non-nil error. If any AckFunc is left uncalled
+	// the connector will not be able to exit gracefully. The WriteAsync
+	// function should only return an error in case a critical error happened,
+	// it is expected that all future calls to WriteAsync will fail with the
+	// same error.
+	// If the plugin caches records before writing them to the destination it
+	// needs to store the AckFunc as well and call it once the record is
+	// written.
+	// AckFunc will panic if it's called more than once.
+	// If WriteAsync returns ErrUnimplemented the SDK will fall back and call
+	// Write instead.
+	WriteAsync(context.Context, Record, AckFunc) error
+
+	// Write receives a Record and is supposed to write the record to the
+	// destination right away. If the function returns nil the record is assumed
+	// to have reached the destination.
+	// WriteAsync takes precedence and will be tried first to write a record. If
+	// WriteAsync is not implemented the SDK will fall back and use Write
+	// instead.
+	Write(context.Context, Record) error
+
+	// Flush signals the plugin it should flush any cached records and call all
+	// outstanding AckFunc functions. No more calls to Write will be issued
+	// after Flush is called.
+	Flush(context.Context) error
+
+	// Teardown signals to the plugin that all records were written and there
+	// will be no more calls to any other function. After Teardown returns, the
+	// plugin should be ready for a graceful shutdown.
+	Teardown(context.Context) error
+
+	mustEmbedUnimplementedDestination()
+}
+
+type AckFunc func(error) error
+
+func NewDestinationPlugin(impl Destination) cpluginv1.DestinationPlugin {
+	if impl == nil {
+		// prevent nil pointers
+		impl = UnimplementedDestination{}
+	}
+	return &destinationPluginAdapter{impl: impl}
+}
+
+type destinationPluginAdapter struct {
+	impl       Destination
+	wgAckFuncs sync.WaitGroup
+}
+
+func (a *destinationPluginAdapter) Configure(ctx context.Context, req cpluginv1.DestinationConfigureRequest) (cpluginv1.DestinationConfigureResponse, error) {
+	err := a.impl.Configure(ctx, req.Config)
+	return cpluginv1.DestinationConfigureResponse{}, err
+}
+
+func (a *destinationPluginAdapter) Start(ctx context.Context, req cpluginv1.DestinationStartRequest) (cpluginv1.DestinationStartResponse, error) {
+	err := a.impl.Open(ctx)
+	return cpluginv1.DestinationStartResponse{}, err
+}
```

Comment:
Regarding the `Start`/`Run` discussion I'll refer you to [my comment](https://github.com/ConduitIO/conduit/pull/109#discussion_r795726653) on #109. TL;DR: the functions are split because one is a synchronous request/response while the other opens a bidirectional stream, we can't combine them (at least not in a nice way).

Regarding `Stop`/`Teardown`, there's a reason for the split here as well. Let's think about the source plugin - it is producing a stream of records and sending them to Conduit. I like to think about this stream as if it was a Go channel, essentially it is very similar - the plugin is the producer and Conduit is the consumer. And the consumer (Conduit) can't really determine when there is nothing more to read, it is the producer (plugin) who knows if there are any records left and when to actually close the stream. That's why we need `Stop`, to tell the plugin to stop producing new records and flush any records that might be cached in memory.

Now, why do we need `Teardown`? Because we have a stream of acknowledgments going back to the plugin, so even after `Stop` is called the ack stream stays open so that Conduit can deliver the acknowledgments for all remaining records. After it delivered all acknowledgments back to the plugin it still can't safely shut down the plugin because processing the acknowledgments might take some time, so that's where `Teardown` comes in - it is a signal to the plugin that Conduit won't send any more acknowledgments and wants to shut it down. The plugin can wait in that function that all acknowledgments are successfully processed and that any cleanup needed is executed, after that the plugin can safely be stopped.

Sidenote: a lot of this complexity is hidden in the SDK so that the plugin developer ideally does not have to think about it.